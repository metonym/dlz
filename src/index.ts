export { npmPackage } from "./npm-package";
