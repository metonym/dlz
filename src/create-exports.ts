import { join, parse, sep } from "node:path";
import { totalist } from "totalist/sync";
import { essential_files } from "./npm-package";

interface ExportsMap {
  [key: string]: {
    types?: string | undefined;
    svelte?: string;
    import?: string;
  };
}

const removeJs = (name: string) => name.replace(/\.js$/, "");

const extTs = (name: string) => removeJs(name) + ".d.ts";

function isIndexFile(name: string): name is `${string}index.js` {
  return parse(name).base === "index.js";
}

function isTsOrTypeDef(name: string): name is `./${string}.ts` {
  return /(.*\.d\.ts$)|(.*\.ts$)/.test(name);
}

function extHasTypes(ext: string): ext is ".js" | ".svelte" {
  return [".js", ".svelte"].includes(ext);
}

function prefixRelative<T extends string>(name: T): `./${T}` {
  return `./${name}`;
}

const getExports = (filename: string) => {
  let file = filename;
  let folder = "";

  if (filename.includes(sep)) {
    folder = parse(filename).dir;
  }

  const is_root = !folder;
  const file_path = prefixRelative(file);

  if (isIndexFile(file)) {
    return {
      [is_root ? "." : prefixRelative(folder)]: {
        types: extTs(file_path),
        import: file_path,
        svelte: is_root ? file_path : undefined,
      },
    };
  } else {
    if (!isTsOrTypeDef(file)) {
      const ext = parse(file).ext;
      const import_path = prefixRelative(join(folder, `*${ext}`));
      const import_path_wildcard = prefixRelative(join(folder, "*"));

      const map = {
        [import_path]: {
          types: extHasTypes(ext) ? extTs(import_path) : undefined,
          import: import_path,
        },
      };

      // To support extension-less imports, each JS file should have an explicit mapping.
      if (ext === ".js") {
        if (is_root) {
          map[removeJs(file_path)] = {
            types: extTs(file_path),
            import: file_path,
          };
        } else {
          map[import_path_wildcard] = {
            types: extHasTypes(ext) ? extTs(import_path) : undefined,
            import: import_path,
          };
        }
      }

      return map;
    }
  }
};

export function createExports(dir: string) {
  const exports_map: ExportsMap = {};

  totalist(dir, (name) => {
    if (essential_files.includes(name)) return;
    Object.assign(exports_map, getExports(name));
  });

  return exports_map;
}
