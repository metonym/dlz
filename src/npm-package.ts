import {
  copyFileSync,
  cpSync,
  existsSync,
  mkdirSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { createExports } from "./create-exports";

const dir = "./package";
const src = "./src";
const pkg_json = join(dir, "package.json");

export function npmPackage() {
  console.time("package");

  if (existsSync(dir)) {
    rmSync(dir, { recursive: true });
  }

  mkdirSync(dir);

  ["./package.json", "./README.md", "./LICENSE"].forEach((file) => {
    copyFileSync(file, join(dir, file));
  });

  cpSync(src, dir, { recursive: true });

  const pkgJson = JSON.parse(readFileSync(pkg_json, "utf8"));

  delete pkgJson.scripts;
  delete pkgJson.devDependencies;

  pkgJson.exports = createExports(dir);

  writeFileSync(pkg_json, JSON.stringify(pkgJson, null, 2));

  console.timeEnd("package");
}
