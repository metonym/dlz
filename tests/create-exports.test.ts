import { expect, test } from "bun:test";
import { createExports } from "../src/create-exports";

test("create-exports", () => {
  expect(createExports("./tests/mocks")).toMatchSnapshot();
});
