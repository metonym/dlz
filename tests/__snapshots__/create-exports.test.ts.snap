// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`create-exports 1`] = `
{
  ".": {
    "import": "./index.js",
    "svelte": "./index.js",
    "types": "./index.d.ts",
  },
  "./*.js": {
    "import": "./*.js",
    "types": "./*.d.ts",
  },
  "./*.svelte": {
    "import": "./*.svelte",
    "types": "./*.svelte.d.ts",
  },
  "./add": {
    "import": "./add.js",
    "types": "./add.d.ts",
  },
  "./nested/*.json": {
    "import": "./nested/*.json",
    "types": undefined,
  },
  "./nested/nested-again": {
    "import": "./nested/nested-again/index.js",
    "svelte": undefined,
    "types": "./nested/nested-again/index.d.ts",
  },
  "./subfolder": {
    "import": "./subfolder/index.js",
    "svelte": undefined,
    "types": "./subfolder/index.d.ts",
  },
  "./subfolder/*": {
    "import": "./subfolder/*.js",
    "types": "./subfolder/*.d.ts",
  },
  "./subfolder/*.css": {
    "import": "./subfolder/*.css",
    "types": undefined,
  },
  "./subfolder/*.js": {
    "import": "./subfolder/*.js",
    "types": "./subfolder/*.d.ts",
  },
  "./subfolder/*.svelte": {
    "import": "./subfolder/*.svelte",
    "types": "./subfolder/*.svelte.d.ts",
  },
}
`;
